feat: get transactions from mem pool ordered by fee

vasild opened the pull request:
## Proposed changes
<!--
Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. If it fixes a bug or resolves a feature request, be sure to link to that issue.
-->

## Types of changes
<!--
What types of changes does your code introduce?
_Put an `x` in the boxes that apply and remove the rest of them: keep this PR as concise, but descriptive, as possible.._
-->

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Refactoring (improve a current implementation without adding a new feature or fixing a bug)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Build (changes that affect the build system)
- [ ] Docs (documentation only changes)
- [ ] Test (adding missing tests or fixing existing tests)
- [ ] Other... Please describe:

## Checklist
<!--
_Put an `x` in the boxes that apply and remove this text and the rest of them. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._
-->

- [x] I have read the [CONTRIBUTING](https://docs.ark.io/developers/guidelines/contributing.html) documentation
- [x] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)
<!--

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...
-->

vasild commented:
Beware:

before this patch:
```
      ✓ multiple additions and retrievals (646ms)
      ✓ delete + add after sync (173ms)
```

after this patch:
```
      ✓ multiple additions and retrievals (5779ms)
      ✓ delete + add after sync (1306ms)
```

vasild commented:
This implements the first part of https://github.com/ArkEcosystem/core/issues/1061

faustbrian commented on packages/core-transaction-pool-mem/lib/mem-pool-transaction.js:
```diff
@@ -0,0 +1,36 @@
+  }
+
+  set sequence (seq) {
+    assert.strictEqual(this.sequence_, undefined)
```
```suggestion
    assert.strictEqual(this._sequence, undefined)
```

vasild replied:
Done

faustbrian commented on packages/core-transaction-pool-mem/lib/mem-pool-transaction.js:
```diff
@@ -0,0 +1,36 @@
+
+  get sequence () {
+    assert.notEqual(this.sequence_, undefined)
+    return this.sequence_
```
```suggestion
    return this._sequence
```

vasild replied:
Done

faustbrian commented on packages/core-transaction-pool-mem/lib/mem-pool-transaction.js:
```diff
@@ -0,0 +1,36 @@
+  }
+
+  get transaction () {
+    return this.transaction_
```
```suggestion
    return this._transaction
```

vasild replied:
Done

faustbrian commented on packages/core-transaction-pool-mem/lib/mem-pool-transaction.js:
```diff
@@ -0,0 +1,36 @@
+
+    if (sequence !== undefined) {
+      assert(sequence instanceof Number)
+      this.sequence_ = sequence
```
```suggestion
      this._sequence = sequence
```

faustbrian commented on packages/core-transaction-pool-mem/lib/mem-pool-transaction.js:
```diff
@@ -0,0 +1,36 @@
+class MemPoolTransaction {
+  constructor (transaction, sequence) {
+    assert(transaction instanceof Transaction)
+    this.transaction_ = transaction
```
```suggestion
    this._transaction = transaction
```

vasild replied:
Done

faustbrian commented on packages/core-transaction-pool-mem/lib/mem-pool-transaction.js:
```diff
@@ -0,0 +1,36 @@
+
+  set sequence (seq) {
+    assert.strictEqual(this.sequence_, undefined)
+    this.sequence_ = seq
```
```suggestion
    this._sequence = seq
```

vasild replied:
Done

faustbrian commented on packages/core-transaction-pool-mem/lib/mem-pool-transaction.js:
```diff
@@ -0,0 +1,36 @@
+ * A mem pool transaction.
+ * A normal transaction + a sequence number used to order by insertion time.
+ */
+class MemPoolTransaction {
```
```suggestion
module.exports = class MemPoolTransaction {
```

vasild replied:
Done

vasild commented:
> Beware:
> 
> before this patch:
> 
> ```
>       ✓ multiple additions and retrievals (646ms)
>       ✓ delete + add after sync (173ms)
> ```
> 
> after this patch:
> 
> ```
>       ✓ multiple additions and retrievals (5779ms)
>       ✓ delete + add after sync (1306ms)
> ```

Just tried to sort on demand every time a sorted result is needed instead of storing the data sorted. We need two orders: by fee (to get highest fee transactions) and by expiration (to expire old transactions). This would make sense if we insert a lot (because then we do not do any sorting during insert) and read a little (because reading is expensive as it involves sorting):

```
      ✓ multiple additions and retrievals (29853ms)
      ✓ delete + add after sync (233ms)
```

Not good either.

vasild commented:
> > Beware:
> > before this patch:
> > ```
> >       ✓ multiple additions and retrievals (646ms)
> >       ✓ delete + add after sync (173ms)
> > ```
> > after this patch:
> > ```
> >       ✓ multiple additions and retrievals (5779ms)
> >       ✓ delete + add after sync (1306ms)
> > ```
> 
> Just tried to sort on demand every time a sorted result is needed instead of storing the data sorted. We need two orders: by fee (to get highest fee transactions) and by expiration (to expire old transactions). This would make sense if we insert a lot (because then we do not do any sorting during insert) and read a little (because reading is expensive as it involves sorting):
> 
> ```
>       ✓ multiple additions and retrievals (29853ms)
>       ✓ delete + add after sync (233ms)
> ```
> 
> Not good either.

Using http://www.collectionsjs.com/sorted-set:
```
      ✓ multiple additions and retrievals (21493ms)
      ✓ delete + add after sync (177ms)
```

vasild commented:
> @vasild what dataset size are you using that you end up with 21 seconds, millions of records or is the 1024 from the tests?

It is 1024 from the tests (unmodified), but this does not matter so much. What matters is the magnitude of the change: before patch <1s, with array, >5s. Simply not acceptable. Also, I would expect the sorted-map solution to be faster than the array, not 4x slower. I have not fully analyzed this, just sharing intermediate results.

vasild commented:
Added one more test and changed the sorting to happen on demand / lazily.

Looks on par with the performance as of before order by fee (before this patch):

```
before:
      ✓ multiple additions and retrievals (635ms)
      ✓ delete + add after sync (173ms)
      ✓ add many then get first few (821ms)

after:
      ✓ multiple additions and retrievals (717ms)
      ✓ delete + add after sync (181ms)
      ✓ add many then get first few (865ms)
```

vasild commented on packages/core-transaction-pool-mem/package.json:
```diff
@@ -28,7 +28,8 @@
     "delay": "^4.1.0",
-    "fs-extra": "^7.0.0"
+    "fs-extra": "^7.0.0",
+    "random-seed": "^0.3.0"
```
Right, good catch! Moved. Thanks!
